juju/juju
# provider/vsphere: cache VMDKs in datastore

## Description of change

Cache VMDKs in a controller-specific location
in the datastore, and use that when creating
VMs. This serves two purposes:

 - to avoid downloading and uploading for each
   new virtual machine
 - to avoid having to upload to the hosts, as
   we were previously doing (uploading to the
   NFC lease location)

We now also avoid downloading the OVA to disk,
instead piping the VMDK contents to the datastore
upload.

The VMDK we upload to the datastore is not stored
in the virtual disk format, so to create a functioning
VM we must:
 1. create a temporary VM with the VMDK attached
 2. clone that VM to create a copy of the VMDK in
   the virtual disk format

We continue to use OVF/ImportVApp for the purpose
of creating the initial VM, which keeps things the
same between vCenter and ESXi. Instead of pulling
the OVF out of the OVA file, we now store an OVF
definition in the source tree and compile it in. This
means we don't have to read the OVA file at all if
the VMDK is already cached.

This PR is based on top of https://github.com/juju/juju/pull/7855, so you can ignore the first commit.

## QA steps

1. juju bootstrap vsphere
2. juju add-machine --constraints root-disk=20G
3. juju add-machine --series=trusty
(ensure all machines come up OK, and check in debug-log that the machine added in step 2 did not incur an upload; ssh to machine from step 2 to check its root disk size is 20GiB)
4. check datastore, it should have VMDKs cached in juju-vmdks/<controller-uuid>/{xenial,trusty}.
5. delete the cache from datastore
6. juju add-machine
(ensure machine comes up OK)
7 juju destroy-controller -y vsphere --destroy-all-models
(ensure datastore cache is deleted automatically)

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1711019
